pkg/objstore/backoff.go

```diff
@@ -0,0 +1,248 @@
+package objstore
+
+import (
+	"context"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/thanos-io/thanos/pkg/runutil"
+)
+
+type backoffBucket struct {
+	Bucket
+	runutil.Backoff
+	logger        log.Logger
+	ignoreObjects []string
+
+	retries *prometheus.CounterVec
+}
+
+func NewBucketWithBackoff(bkt Bucket, backoff runutil.Backoff, logger log.Logger, reg prometheus.Registerer, ignoreObjects ...string) InstrumentedBucket {
+	b := &backoffBucket{
+		Bucket:        bkt,
+		Backoff:       backoff,
+		logger:        logger,
+		ignoreObjects: ignoreObjects,
+		retries: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
```

Comment:
An histogram may be more useful here. If you have a buckets from 0 util the max number of retries, then you could track the number of retries for each API call in order to have a better split.